File: pkg-py/src/shinychat/_html_islands.py
from __future__ import annotations

from dataclasses import dataclass
from itertools import groupby
from typing import Union

from htmltools import (
    HTML,
    HTMLDependency,
    Tag,
    TagChild,
    Tagifiable,
    TagifiedTag,
    TagifiedTagList,
    TagList,
)


def _resolve_tagifiable(content: TagChild) -> TagChild:
    """Resolve a Tagifiable to its Tag form (if it isn't already a Tag/TagList/str)."""
    if isinstance(content, (Tag, TagifiedTag, TagList, TagifiedTagList, str)):
        return content
    if isinstance(content, Tagifiable):
        return content.tagify()
    return content


def _has_react_attr(child: TagChild) -> bool:
    """Check if a tag child has the data-shinychat-react attribute."""
    child = _resolve_tagifiable(child)
    if isinstance(child, (Tag, TagifiedTag)):
        return "data-shinychat-react" in child.attrs
    return False


def split_content_by_trust(
    content: TagChild | TagList,
) -> list[tuple[bool, TagChild]]:
    """Split mixed content into trusted and untrusted runs.

    Plain strings may contain model output and are untrusted. HTML()-marked
    strings, tags, and Tagifiable values are server-authored UI and trusted.
    """
    if isinstance(content, (TagList, TagifiedTagList)):
        children = list(content)
    else:
        children = [content]

    def is_trusted(child: TagChild) -> bool:
        return not isinstance(child, str)

    result: list[tuple[bool, TagChild]] = []
    for trusted, group_iter in groupby(children, is_trusted):
        group = list(group_iter)
        if trusted:
            result.append((True, TagList(*group)))
        else:
            result.append((False, "".join(str(child) for child in group)))
    return result


@dataclass
class IslandBlockPart:
    """Rendered HTML and dependencies from a run of trusted, non-React content."""

    html: str
    deps: list[HTMLDependency]


@dataclass
class IslandResidualPart:
    """Rendered HTML and dependencies from a run of bare ``data-shinychat-react`` elements."""

    html: str
    deps: list[HTMLDependency]


IslandPart = Union[IslandBlockPart, IslandResidualPart]


def derive_island_parts(content: TagChild | TagList) -> list[IslandPart]:
    """Split trusted tag content around ``data-shinychat-react`` elements.

    Plain strings are not accepted. They are markdown and must be handled
    by the caller. This function raises ``TypeError`` for them.

    Runs without the attribute become :class:`IslandBlockPart` parts
    (trusted HTML and deps for ``html_block`` envelopes). Elements with
    the attribute become :class:`IslandResidualPart` string runs.
    :class:`ChatMessage` and :class:`MarkdownStream` share this one
    derivation. No wrapper tag is constructed.
    """
    if isinstance(content, str) and not isinstance(content, HTML):
        raise TypeError(
            "derive_island_parts() requires trusted tag content; plain "
            "strings are markdown and must be handled by the caller."
        )
    if isinstance(content, (TagList, TagifiedTagList)):
        children: list[TagChild] = list(content)
    else:
        children = [content]

    parts: list[IslandPart] = []
    for is_react, group in groupby(children, _has_react_attr):
        if is_react:
            for item in group:
                rendered = TagList(item).render()
                run = f"\n\n{rendered['html']}\n\n"
                if parts and isinstance(parts[-1], IslandResidualPart):
                    parts[-1].html += run
                    parts[-1].deps.extend(rendered["dependencies"])
                else:
                    parts.append(
                        IslandResidualPart(
                            html=run, deps=list(rendered["dependencies"])
                        )
                    )
        else:
            island = TagList(*group).render()
            parts.append(
                IslandBlockPart(
                    html=island["html"], deps=list(island["dependencies"])
                )
            )
    return parts

File: pkg-r/R/html_islands.R
#' Split tag content around elements with data-shinychat-react
#'
#' Elements WITH the attribute are emitted bare.
#' Consecutive elements WITHOUT the attribute are grouped into
#' <shiny-chat-raw-html> wrappers.
#'
#' @param content A tag, tagList, or other HTML content.
#' @return A list of tag children ready to be serialized.
#' @noRd
split_html_islands <- function(content) {
  # Convert to tags so custom classes (e.g., shinychat_tool_card)
  # resolve their data-shinychat-react attribute
  content <- htmltools::as.tags(content)

  if (inherits(content, "shiny.tag")) {
    if (has_react_attr(content)) {
      return(list(content))
    }
    return(list(htmltools::tag("shiny-chat-raw-html", list(content))))
  }

  if (inherits(content, "shiny.tag.list")) {
    children <- as.list(content)
  } else {
    return(list(htmltools::tag("shiny-chat-raw-html", list(content))))
  }

  if (length(children) == 0) {
    return(list())
  }

  is_react <- vapply(children, has_react_attr, logical(1))
  group_id <- cumsum(c(TRUE, diff(is_react) != 0))
  groups <- split(children, group_id)

  result <- list()
  for (group in groups) {
    if (has_react_attr(group[[1]])) {
      result <- c(result, group)
    } else {
      result <- c(result, list(htmltools::tag("shiny-chat-raw-html", group)))
    }
  }
  result
}

#' Split mixed content into ordered provenance runs
#'
#' Plain character values may contain model output and are untrusted.
#' HTML()-marked strings and tags are server-authored UI and trusted.
#'
#' @param content Content accepted by htmltools.
#' @return A list of lists containing `trusted` and `content`.
#' @noRd
split_content_by_trust <- function(content) {
  if (inherits(content, "shiny.tag.list")) {
    children <- as.list(content)
  } else {
    children <- list(content)
  }

  is_trusted <- vapply(
    children,
    function(child) !(is.character(child) && !inherits(child, "html")),
    logical(1)
  )
  if (length(children) == 0) {
    return(list(list(trusted = FALSE, content = "")))
  }

  group_id <- cumsum(c(TRUE, diff(is_trusted) != 0))
  groups <- split(children, group_id)
  unname(lapply(groups, function(group) {
    trusted <- !(is.character(group[[1]]) && !inherits(group[[1]], "html"))
    list(
      trusted = trusted,
      content = if (trusted) {
        do.call(htmltools::tagList, group)
      } else {
        paste0(unlist(group, use.names = FALSE), collapse = "")
      }
    )
  }))
}

has_react_attr <- function(child) {
  if (!inherits(child, "shiny.tag")) {
    return(FALSE)
  }
  !is.null(htmltools::tagGetAttribute(child, "data-shinychat-react"))
}
